### ISSUE TITLE
error: "unable to set terminal parameters" on IOS devices

### ISSUE THREAD

**Issue by skregas** [ansible/ansible#46185]
##### SUMMARY

Consistently encountering **AnsibleConnectionFailure: unable to set terminal parameters**
when changing from 
**connection: local**
 to 
**connection: network_cli**
and defining
`ansible_network_os: ios`
for the host.

Also tried adding
`become: true
become_method: enable
`
to the play with no effect
##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

ios_command

##### ANSIBLE VERSION

```
ansible 2.6.4
  config file = /var/lib/awx/ansible.cfg
  configured module search path = [u'/var/lib/awx/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, May 31 2018, 09:41:32) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]

```

##### CONFIGURATION

```

ANSIBLE_COW_SELECTION(/var/lib/awx/ansible.cfg) = random
HOST_KEY_CHECKING(/var/lib/awx/ansible.cfg) = False
PERSISTENT_CONNECT_TIMEOUT(/var/lib/awx/ansible.cfg) = 30

```

##### OS / ENVIRONMENT

Ansbile controller: Red Hat 4.8.5
Target OS: C800, C3900 

##### STEPS TO REPRODUCE

Using **connection: network_cli**
issue a single command using the ios_command module

In inventory, define:
`ansible_network_os: ios`

```
hosts: <inventory>
  gather_facts: no
  connection: network_cli

  tasks:
    - name: Run show version
    ios_command:
      commands: show run | inc logging host
    register: cmd_output
```
Also tried:
```
hosts: <inventory>
  gather_facts: no
  connection: network_cli
  **become: true** 
  **become_method: enable**

  tasks:
    - name: Run show version
    ios_command:
      commands: show run | inc logging host
    register: cmd_output
```

##### EXPECTED RESULTS

The task and playbook complete successfully

##### ACTUAL RESULTS

```
-bash-4.2$ ansible all -i _hostname_, -c network_cli -u fc.acsansible.ro -k -m ios_command -e ansible_network_os=ios -vvvv
ansible 2.6.4
  config file = /var/lib/awx/ansible.cfg
  configured module search path = [u'/var/lib/awx/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, May 31 2018, 09:41:32) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
Using /var/lib/awx/ansible.cfg as config file
SSH password:
setting up inventory plugins
Parsed _hostname_, inventory source with host_list plugin
Loading callback plugin minimal of type stdout, v2.0 from /usr/lib/python2.7/site-packages/ansible/plugins/callback/minimal.pyc
META: ran handlers
<_hostname_> attempting to start connection
<_hostname_> using connection plugin network_cli
<_hostname_> local domain socket does not exist, starting it
<_hostname_> control socket path is /var/lib/awx/.ansible/pc/39555cd441
<_hostname_>
The full traceback is:
Traceback (most recent call last):
  File "/bin/ansible-connection", line 105, in start
    self.connection._connect()
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/network_cli.py", line 328, in _connect
    self._terminal.on_open_shell()
  File "/usr/lib/python2.7/site-packages/ansible/plugins/terminal/ios.py", line 58, in on_open_shell
    raise AnsibleConnectionFailure('unable to set terminal parameters')
AnsibleConnectionFailure: unable to set terminal parameters

_hostname_ | FAILED! => {
    "msg": "unable to set terminal parameters"
}

```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/network/ios/ios_command.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/ios/ios_command.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @rcarrillocruz
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
Hi @skregas,

Thank you for the issue, just so you are aware we have a dedicated Working Group for network.
You can find other people interested in this in #ansible-network on Freenode IRC
For more information about communities, meetings and agendas see https://github.com/ansible/community

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by akira6592** [comment]
Is the privilege level of user `fc.acsansible.ro` `0` ? 
Then, it looks like the following issue.
https://github.com/ansible/ansible/issues/44672

**Comment by skregas** [comment]
@akira6592 privilege level is 15
`#show privilege
Current privilege level is 15
`

**Comment by akira6592** [comment]
Well, can you execute the `terminal length` and `terminal width 512` statements by manually login as `fc.acsansible.ro` user?

**Comment by skregas** [comment]
Can execute `terminal length`, but not `terminal width`. The latter command results in `Command authentication failed`.
So that points to a permissions issue with the `fc.acsansible.ro` account.

The question then becomes: has `ios_command` always behaved this way, needing to set terminal length and width parameters? In Ansible 2.4, the combination of the `fc.acsansible.ro` user executing ios_command never caused this issue. What, if anything, has changed in that module that requires different permissions, or requires setting the terminal length and width?

As stated in the problem description, I've tried adjusting the permissions using `become` and `become_method`. 
Is there any other settings in Ansible I can adjust to effectively get the same behavior from `ios_command` as seen in 2.4?

**Comment by salman1485** [comment]
So, it is fair to assume that starting 2.5 ios_command first tries to run "terminal length 0" and "terminal width 512" and if account used to run ios_command does not have this access then playbooks will break. 

If so then why is this so? Purpose of ios_command was to run random commands on remote ios device. It should not run any commands that can make changes to remote device even if they are minimal

**Comment by pabelanger** [comment]
It sounds like this might be a regression between 2.4 and 2.6, since this seems to work properly on 2.4.  I'd have to look into the code to see if anything changed.

**Comment by privateip** [comment]
@skregas 

> What, if anything, has changed in that module that requires different permissions, or requires setting the terminal length and width?

Calling `terminal length` has always existed.  It is absolutely required for any code that is doing interactions with IOS CLI (not just Ansible).  The use of `terminal width` came later due to a bug where by command strings would exceed the width of the current terminal setting.  Its entirely plausible that this could be turned into a configurable setting.  /cc @Qalthos 

@salman1485 

> So, it is fair to assume that starting 2.5 ios_command first tries to run "terminal length 0" and "terminal width 512" and if account used to run ios_command does not have this access then playbooks will break.

Yes, that is true although there should be a more gracefully message returned to stdout as opposed to the stack trace /cc @Qalthos 

> If so then why is this so? Purpose of ios_command was to run random commands on remote ios device. It should not run any commands that can make changes to remote device even if they are minimal

Setting `terminal length` is absolutely required in order for the connection plugin to work properly with the CLI'  Setting `terminal width` is debatable.   Neither of these settings change the device configuration.  Both just simply set values for the _current_ terminal session.  If you have command authorization enabled, then you will need to give the user that is used by the playbook authorization to run those two terminal commands otherwise Ansible will fail.

**Comment by skregas** [comment]
@pabelanger I went through the code that's using `terminal width` and `terminal length`. It hasn't changed between 2.4 and 2.6.

However, we downgraded Ansible from 2,6 to 2.4, ran the same playbook successfully to completion. Nothing else changed in the file. 

[Results from 2.4](https://gist.github.com/skregas/3021372e51d43e9a6f654fcfada15acc)

Results from 2.6 are detailed at the top of this issue

@privateip For tests in both 2.4 and 2.6, command authorization is not enabled. The test is just a plain ios_command call .
